prov/shm/src/smr_fabric.c

```diff
@@ -54,6 +54,7 @@ static int smr_fabric_close(fid_t fid)
 	if (ret)
 		return ret;
 	free(fabric);
+	smr_cleanup();
```

Comment:
If the user opens 2 fabric instances, doesn't this result in closing all EPs when the first instance is closed?

Reply:
If by "closing all EPs" you mean those EPs' names won't be list, then yes. The EPs will still exist, of course. I think we might need some sort of id tag for the ep names so that only the ep names that belong to that fabric are cleaned up. I don't know how to avoid this otherwise.

Reply:
Why not call smr_cleanup() from fini, like the previous code did?  Anyone using shm directly would need to call smr_cleanup().

The fabric instance should have ref counts on it by the domains, which should be referenced by the endpoints.  So in order to close the fabric here, we already should have closed and cleaned up all related endpoints.  I thought he cleanup in fini was to handle the case where the app wasn't behaving and simply exited without cleaning up.

Reply:
I thought you said the DL builds did not call fini? Or did I misunderstand and it's just INI that it doesn't call?

Reply:
The DL builds do not call INI or FINI.  However, a DL build that uses the shm_util code can call smr_cleanup().  They would need something like this:
```
#if HAVE_EFA_DL
    smr_cleanup();
# 
```
This is similar to how it needs to handle initialization.  See tcpx_init.c for an example.

Reply:
Yeah, the intent was to do the cleanup at a process scope outside the lifecycle of the libfabric objects, so fini seems like the right place. We can have a CFLAG check for DSO builds like how EFA and verbs does in their ini functions to initialize some objects.

Reply:
Jinx.

Reply:
Got it. Updated.